### PR TITLE
USCensusPEP_PopulationEstimatebyRace

### PR DESCRIPTION
**Pull request by Harsha-chandaluri**

### CODE REVIEW

File: scripts/us_census/pep/population_estimate_by_race/README.md [hunk 1]
```diff
@@ -50,11 +50,31 @@ Run the test cases
...
+[Updated the script on November 06, 2024]
+Downloading input files is now integrated into preprocess.py, eliminating the need to run the separate download.sh script. 
+All source file URLs, including future URLs adhering to the same structure, are centrally managed in the input_url.json file.
+All input files required for processing should be stored within the designated "input_files" folder.
+
+
+### Downloading and Processing Data
+
+To perform "download and process", run the below command:
+`python3 preprocess.py --mode=""`
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/population_estimate_by_race/README.md:62]: remove "--mode=""` in this line

File: scripts/us_census/pep/population_estimate_by_race/README.md [hunk 2]
```diff
@@ -50,11 +50,31 @@ Run the test cases
...
+
+Running this command generates input_fles and csv, mcf, tmcf files
+
+
+   If you want to perform "only process", run the below command:
+
+        `python3 preprocess.py --mode="process"`
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/population_estimate_by_race/README.md:69]: double quotes not required in the mode

File: scripts/us_census/pep/population_estimate_by_race/README.md [hunk 3]
```diff
@@ -50,11 +50,31 @@ Run the test cases
...
+        
+   If you want to perform "only download", run the below command:
+
+        `python3 preprocess.py --mode="download"`
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/population_estimate_by_race/README.md:73]: double quotes not required in the mode

File: scripts/us_census/pep/population_estimate_by_race/preprocess_test.py [hunk 4]
```diff
@@ -36,13 +36,13 @@ class TestPreprocess(unittest.TestCase):
     mcf_file_path = OP_DATA_FOLDER
     tmcf_file_path = OP_DATA_FOLDER
 
-    ip_files = os.listdir(os.path.join(TEST_DATA_FOLDER, "test_files"))
-    ip_data_path = [
-        os.path.join(os.path.join(TEST_DATA_FOLDER, "test_files"), file_name)
-        for file_name in ip_files
-    ]
+    ip_files = os.path.join(TEST_DATA_FOLDER, "test_files")
+    # ip_data_path = [
```
Comment (krishnaswamypradeep) [inline scripts/us_census/pep/population_estimate_by_race/preprocess_test.py:40]: Remove these commented lines

**Review by krishnaswamypradeep — COMMENTED** [review]
Please fix the comments.

File: scripts/us_census/pep/population_estimate_by_race/preprocess_test.py (hunk 4 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess_test.py]: Removed

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 5]
```diff
@@ -951,18 +1128,173 @@ def _generate_tmcf(self, df_cols: list, name: str) -> None:
...
+                is_file_downloaded = True
+
+            except Exception as e:
+                logging.error(f"Retry file download {_url} - {e}")
+                time.sleep(5)
+                retry_number += 1
+                if retry_number > max_retry:
+                    logging.error(f"Error downloading {_url}")
+                    logging.error("Exit from script")
+                    sys.exit(0)
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1266]: for failures, exit with a non-zero status or use logging.fatal()

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 6]
```diff
@@ -951,18 +1128,173 @@ def _generate_tmcf(self, df_cols: list, name: str) -> None:
...
                                   "output")
+    cleaned_csv_path = data_file_path
     mcf_path = data_file_path
     tmcf_path = data_file_path
-    loader = CensusUSAPopulationByRace(ip_files, data_file_path, mcf_path,
-                                       tmcf_path)
-    loader.process()
+    input_path = _FLAGS.input_path
+
+    if mode == "":
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1280]:
```
    if mode == "" or mode == "download":
        # download & process
        add_future_year_urls()
        download_status = download_files()
   if (mode == "" and download_status) or mode == "process":
            loader = CensusUSAPopulationByRace(input_path, data_file_path,
                                               mcf_path, tmcf_path)
            loader.process()
```

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 7]
```diff
@@ -760,103 +891,149 @@ def _transform_data(self, df: pd.DataFrame) -> None:
...
+
+            name = "USA_Population_Count_by_Race_National_state_2000"
+            generator_df=pd.read_csv\
+                (os.path.join(self.cleaned_csv_file_path,
+                "USA_Population_Count_by_Race_National_state_2000.csv"))
+
+            self._generate_mcf(generator_df.columns, name)
+            self._generate_tmcf(generator_df.columns, name)
+        else:
+            logging.error(
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1034]: logging.fatal() or sys.exit(1) to return error status code on failure

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 8]
```diff
@@ -760,103 +891,149 @@ def _transform_data(self, df: pd.DataFrame) -> None:
...
+            if result:
+                processed_count += 1
+        logging.info(f"No of files processed {processed_count}")
+        if processed_count == total_files_to_process & total_files_to_process > 0:
+
+            name = "USA_Population_Count_by_Race_before_2000"
+            generator_df=pd.read_csv\
+                (os.path.join(self.cleaned_csv_file_path,
+                    "USA_Population_Count_by_Race_before_2000.csv"))
+            #print("Before delete ", generator_df.shape)
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1009]: Can use logging.debug("...") for debug statements.

They will be shown by dong the following in main()
  logging.verbosity(1)

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 9]
```diff
@@ -760,103 +891,149 @@ def _transform_data(self, df: pd.DataFrame) -> None:
         Returns:
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:891]: Pls update the function block comments to reflect new arguments.

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 10]
```diff
@@ -760,103 +891,149 @@ def _transform_data(self, df: pd.DataFrame) -> None:
...
+                                        index=False)
+            #Added by Shamim to resolve 2020 inconsistent data remove
+            df_national_state_2000 = df_national_state_2000.drop_duplicates(
+                subset=['geo_ID', 'Year'], keep='last')
+            df_national_state_2000.to_csv(os.path.join(
+                self.cleaned_csv_file_path,
+                "USA_Population_Count_by_Race_National_state_2000.csv"),
+                                          index=False)
+        except Exception as e:
+            logging.error("error processing file")
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:976]: pls log the exception when catching it with more context:
```logging.error(f'Error when processing file: {file}: {e}')

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 11]
```diff
@@ -760,103 +891,149 @@ def _transform_data(self, df: pd.DataFrame) -> None:
...
-
-        name = "USA_Population_Count_by_Race_National_state_2000"
-        generator_df=pd.read_csv\
-            (os.path.join(self.cleaned_csv_file_path,
-            "USA_Population_Count_by_Race_National_state_2000.csv"))
-        self._generate_mcf(generator_df.columns, name)
-        self._generate_tmcf(generator_df.columns, name)
+            result = self._transform_data(df)
+            if result:
+                processed_count += 1
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1001]: for fialures, pls add a logging.error:
```logging.error(f'Failed to process {file}')```

**Review by krishnaswamypradeep — APPROVED** [review]
Thanks Harsha. Looks good

File: scripts/us_census/pep/population_estimate_by_race/download.py [hunk 12]
```diff
@@ -105,7 +105,7 @@ def _save_data(_url: str, download_local_path: str) -> None:
                 df = pd.read_csv(_url, on_bad_lines='skip', names=cols)
             df.to_excel(download_local_path + os.sep + file_name,\
                 index=False,engine='xlsxwriter')
-        elif "co-est00int-alldata" in _url or "CC-EST2020-ALLDATA" in _url:
+        elif "co-est00int-alldata" in _url or "CC-EST2020-ALLDATA" in _url or "cc-est2022-all":
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/download.py:108]: last condition will always make this condition true.

did you mean `or "cc-est2022-all" in _url`

File: scripts/us_census/pep/population_estimate_by_race/download.py [hunk 13]
```diff
@@ -105,7 +105,7 @@ def _save_data(_url: str, download_local_path: str) -> None:
...
             df = pd.read_csv(_url,
                              on_bad_lines='skip',
                              encoding='ISO-8859-1',
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/download.py:111]: pls add an error if none of the above 
```
    else:
       logging.error(f'Unknown file')
```

File: scripts/us_census/pep/population_estimate_by_race/download.py [hunk 14]
```diff
@@ -105,7 +105,7 @@ def _save_data(_url: str, download_local_path: str) -> None:
                 df = pd.read_csv(_url, on_bad_lines='skip', names=cols)
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/download.py:105]: in line 82 above, pls add context when printing:
https://github.com/datacommonsorg/data/pull/1105/files#diff-cdbbb40fb98f2f2fcc3c38625b8d29b98cc6eac339b4f3d6a3e73e78e9241b37R82

```
print(f'Downloading {file_name} from {_url}')
```

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 15]
```diff
@@ -174,117 +201,134 @@ def _clean_county_80_xls_file(df: pd.DataFrame, file_path: str) -> pd.DataFrame:
...
+        df['FIPS'] = [f'{x:05}' for x in df['FIPS']]
+        df['Info'] = df['Year'].astype(str) + '-' + df['FIPS'].astype(str)
+        df.drop(columns=['Year','FIPS',1,2,3,4,5,6,7,8,9,10,11,12,13,14\
+            ,15,16,17,18],inplace=True)
+        df = df.groupby(['Info','Race/Sex']).sum().transpose().\
+            stack(0).reset_index()
+        df['Year'] = df['Info'].str.split('-', expand=True)[0]
+        df['geo_ID'] = "geoId/" + df['Info'].str.split('-', expand=True)[1]
+        # Deriving the total values as per the requires SV's
+        df['Total']=df['White male']+df['White female']+df['Black male']\
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:217]: Do we use 'Total' in the output? If so, can we add 'dcAggregate/' prefix to the measurementMethod?

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 16]
```diff
@@ -174,117 +201,134 @@ def _clean_county_80_xls_file(df: pd.DataFrame, file_path: str) -> pd.DataFrame:
...
+        df = df.groupby(['Year','geo_ID']).sum().\
+            reset_index()
+        df.drop(columns=['Total'], inplace=True)
+        df_temp = pd.DataFrame()
+        df_temp = pd.concat([df, df_temp])
+        df_temp['geo_ID'] = "country/USA"
+        df_temp = df_temp.groupby(['Year','geo_ID']).sum().\
+            reset_index()
+        final_df = pd.concat([final_df, df, df_temp])
+        final_df = final_df.drop_duplicates()
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:236]: Is there any duplicate observation without the drop_duplicates()?
Is striping geoId to the last 3 digits in line 226: `df['geo_ID'] = df['geo_ID'].str[:-3]` leading to rows with same geoId although the source didn't have any?

If so, pls add a comment on what types of dups are dropped?

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 17]
```diff
@@ -702,17 +874,21 @@ def _load_data(self, file: str) -> pd.DataFrame:
                 df = _clean_xls_file(df, file)
         elif ".csv" in file:
             if "srh" in file:
+                # csv_data = []
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:877]: pls delete unused code

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 15 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:217]: No we don't use "Total" in output 
This is for SV level total in the code 
prefix to the measurementMethod is not required here

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 18]
```diff
@@ -724,14 +900,34 @@ def _load_data(self, file: str) -> pd.DataFrame:
                     df['geo_ID'].map(str)).str[:len('geoId/NN')]
                 df_state = df_state.groupby(['Year', 'geo_ID']).sum().\
                     reset_index()
-                df = df.append(df_state, ignore_index=True)
-                df = df.append(df_national, ignore_index=True)
+                df = pd.concat([df, df_state], ignore_index=True)
+                df = pd.concat([df, df_national], ignore_index=True)
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:904]: concat in a single statement is faster:
`df = pd.concat([df, df_state, df_national], ignore_index=True)`

Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:904]: modified as suggested

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 17 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:877]: Removed all the unused code

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 19]
```diff
@@ -745,118 +941,163 @@ def _load_data(self, file: str) -> pd.DataFrame:
...
-                                      index=False)
+                "Count_Person_TwoOrMoreRaces","Count_Person_NonWhite"]]
+            df_before_2000 = self.df[self.df["Year"] < 2000]
+            df_county_after_2000 = self.df[(self.df["Year"] >= 2000) &
+                                           (self.df["geo_ID"] != "country/USA")
+                                           & (self.df["geo_ID"].str.len() > 9)]
+            df_national_state_2000 = self.df[(self.df["Year"] >= 2000) & (
+                (self.df["geo_ID"].str.len() <= 9) |
+                (self.df["geo_ID"] == "country/USA"))]
+            df_before_2000 = df_before_2000.drop_duplicates(
```
[the remark this answers is not in the document]
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1019]: I will analyze and update you shortly

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 20]
```diff
@@ -951,18 +1193,182 @@ def _generate_tmcf(self, df_cols: list, name: str) -> None:
...
+                        if "pe-02" in _url:
+                            df = pd.read_csv(_url, skiprows=7, on_bad_lines='skip', \
+                                names=cols)
+                        else:
+                            df = pd.read_csv(_url,
+                                             on_bad_lines='skip',
+                                             names=cols)
+                        df.to_excel(download_local_path + os.sep + file_name,\
+                            index=False,engine='xlsxwriter')
+                    elif "co-est00int-alldata" in _url or "CC-EST2020-ALLDATA" in _url or "cc-est2022-all":
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1307]: ` or "cc-est2022-all":` will cause the condition to always succeed.

should it be:
`elif "co-est00int-alldata" in _url or "CC-EST2020-ALLDATA" in _url or "cc-est2022-all" in _url:`

File: scripts/us_census/pep/population_estimate_by_race/download.py [hunk 21]
```diff
@@ -144,8 +144,16 @@ def _download(download_path: str, file_urls: list) -> None:
     """
     if not os.path.exists(download_path):
         os.mkdir(download_path)
+    all_files = os.listdir(download_path)
     for _url in file_urls:
```
[the remark this answers is not in the document]
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py:148]: We are no longer using this file

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 12 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py:108]: yes, Modified

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 13 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py:111]: Modified

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 14 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py:105]: we are no longer using this download.py file should I delete this file ?

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 22]
```diff
@@ -15,32 +15,54 @@
...
+from datetime import datetime as dt
+from absl import logging
 # To import util.alpha2_to_dcid
 _COMMON_PATH = os.path.abspath(
     os.path.join(os.path.dirname(__file__), '../../../../'))
 sys.path.insert(1, _COMMON_PATH)
+import warnings
+
+warnings.filterwarnings('ignore')
+#warnings.simplefilter(action='ignore', category=FutureWarning)
```
[the remark this answers is not in the document]
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:40]: Removed

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 20 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1307]: yes missed you add "cc-est2022-all" in _url now its modified

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 16 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:236]: Sure,  I  need some time to analyze  this will update

Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:236]: https://screenshot.googleplex.com/6M7JH8sMC2aaEYh
Hi Ajai there seems to be no duplicate in output files please refer to this screenshot
Shall I go ahead and remove this drop_duplicate line?

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1019]: https://screenshot.googleplex.com/6M7JH8sMC2aaEYh
Hi Ajai there seems to be no duplicate in output files please refer to this screenshot
Shall I go ahead and remove this drop_duplicate line?

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 12 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py:108]: we are not using this download.py file. Should we delete this file ?

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 16 above)
Comment (ajaits) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:235]: yes, lets remove drop_duplicates().

It should fail dc-import lint if there are duplicates in the future.

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 14 above)
Comment (ajaits) [reply scripts/us_census/pep/population_estimate_by_race/download.py:105]: yes lets delete this

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1013]: Yes, pls remove all drop_duplicates()

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 23]
```diff
@@ -951,18 +1187,184 @@ def _generate_tmcf(self, df_cols: list, name: str) -> None:
...
+                                             on_bad_lines='skip',
+                                             names=cols)
+                        df.to_excel(download_local_path + os.sep + file_name,\
+                            index=False,engine='xlsxwriter')
+                    elif "co-est00int-alldata" in url or "CC-EST2020-ALLDATA" in url or "cc-est2022-all" in url:
+                        df = pd.read_csv(url,
+                                         on_bad_lines='skip',
+                                         encoding='ISO-8859-1',
+                                         low_memory=False)
+                        df.to_csv(download_local_path + os.sep + file_name,
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1306]:
```
 else:
    logging.fatal(f'Unknown csv file: {url}')
```

**Review by ajaits — APPROVED** [review]

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 23 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1305]: updated

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 16 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:235]: Removed

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1012]: Removed

File: scripts/us_census/pep/population_estimate_by_race/download.py (hunk 14 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/download.py]: Deleted

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py [hunk 24]
```diff
@@ -1010,22 +1009,14 @@ def _transform_data(self, df: pd.DataFrame, file_path: str) -> None:
...
                 self.cleaned_csv_file_path,
                 "USA_Population_Count_by_Race_before_2000.csv"),
                                   index=False)
-            #Added to resolve 2020 inconsistent data remove
-            df_county_after_2000 = df_county_after_2000.drop_duplicates(
-                subset=['geo_ID', 'Year'], keep='last')
             df_county_after_2000.to_csv(os.path.join(
                 self.cleaned_csv_file_path,
                 "USA_Population_Count_by_Race_county_after_2000.csv"),
                                         index=False)
```
Comment (ajaits) [inline scripts/us_census/pep/population_estimate_by_race/preprocess.py:1019]: Thanks. Line 285 has a drop_duplicates(). Can we remove that too?

**Review by ajaits — APPROVED** [review]

File: scripts/us_census/pep/population_estimate_by_race/preprocess.py (hunk 24 above)
Comment (Harsha-chandaluri) [reply scripts/us_census/pep/population_estimate_by_race/preprocess.py:1019]: Removed

**Pull request merged**